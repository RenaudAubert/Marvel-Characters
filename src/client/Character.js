import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

export default class Character extends Component {
  constructor(props) {
    super(props);

    this.favClick = this.favClick.bind(this);
  }

  favClick() {
    const { character, onFavClicked } = this.props;
    onFavClicked(character.id);
  }

  // Render each character card with image, name and description
  render() {
    const { character, isFav } = this.props;
    const name = character.name || '';
    const description = character.description || 'No description available for this character';
    const imgURL = character.thumbnail && `${character.thumbnail.path}/landscape_large.${character.thumbnail.extension}`;

    return (
      <div className="h-100 card mb-4 box-shadow">
        <Link to={`/characters/${character.id}`}>
          <img className="card-img-top" src={imgURL} alt="" />
        </Link>
        <div className="card-body">
          <h5 className="card-title">{name}</h5>
          <p className={`card-text ${character.description}` ? '' : ' text-danger'}>{description}</p>
        </div>
        <div className="card-footer text-muted">
          <div className="d-flex justify-content-between align-items-center">
            <div className="btn-group">
              <Link to={`/characters/${character.id}`}>
                <button type="button" className="btn btn-sm btn-outline-secondary">
                  <i className="fas fa-search" />
                </button>
              </Link>
              <button type="button" className="btn btn-sm btn-outline-secondary" onClick={this.favClick}>
                <i className={isFav ? 'fas fa-star' : 'far fa-star'} />
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Character.propTypes = {
  character: PropTypes.shape({
    id: PropTypes.number,
    description: PropTypes.string,
    name: PropTypes.string,
    thumbnail: PropTypes.object,
  }).isRequired,
  isFav: PropTypes.bool,
  onFavClicked: PropTypes.func.isRequired
};

Character.defaultProps = {
  isFav: false
};
