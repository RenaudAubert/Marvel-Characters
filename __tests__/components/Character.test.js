import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import apiResult from '../__mocks__/characters';

// Component to be tested
import Character from '../../src/client/Character';

const character = apiResult.data.results[0];

describe('<Character />', () => {
  const mockOnClick = jest.fn();

  describe('return()', () => {
    test('renders the Character component', () => {
      const wrapper = shallow(<Character character={character} onFavClicked={mockOnClick} />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('renders the Character component without description', () => {
      let character2 = character;
      character2.description = "";
      const wrapper = shallow(<Character character={character2} onFavClicked={mockOnClick} />);
      expect(wrapper.exists('.text-danger')).toEqual(true);
    });
  });
});
