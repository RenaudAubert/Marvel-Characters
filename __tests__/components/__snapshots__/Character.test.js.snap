// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Character /> return() renders the Character component 1`] = `
<div
  className="h-100 card mb-4 box-shadow"
>
  <Link
    replace={false}
    to="/1011334"
  >
    <img
      alt=""
      className="card-img-top"
      src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784/landscape_large.jpg"
    />
  </Link>
  <div
    className="card-body"
  >
    <h5
      className="card-title"
    >
      3-D Man
    </h5>
    <p
      className="card-text text-danger"
    />
  </div>
  <div
    className="card-footer text-muted"
  >
    <div
      className="d-flex justify-content-between align-items-center"
    >
      <div
        className="btn-group"
      >
        <Link
          className="text-dark"
          replace={false}
          to="/1011334"
        >
          <i
            className="fas fa-search fa-lg align-bottom"
          />
        </Link>
        <FavButton
          characterId={1011334}
          isFav={false}
          onFavClicked={[MockFunction]}
        />
      </div>
    </div>
  </div>
</div>
`;
